python/ray/dataframe/dataframe.py

```diff
@@ -1162,79 +1401,149 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
                   .format(expecting=expecting, method=method)
             raise ValueError(msg)
 
-        partition_idx = [
-            self._index.loc[
-                self._index['partition'] == i
-            ].index
-            for i in range(len(self._df))
-        ]
-
-        def fillna_part(df, real_index):
-            old_index = df.index
-            df.index = real_index
-            new_df = df.fillna(value=value, method=method, axis=axis,
-                               limit=limit, downcast=downcast, **kwargs)
-            new_df.index = old_index
-            return new_df
+        if inplace:
+            new_obj = self
+        else:
+            new_obj = self.copy()
 
-        new_df = [
-            _deploy_func.remote(
-                fillna_part,
-                part, partition_idx[i]
-            )
-            for i, part in enumerate(self._df)
-        ]
-
-        new_df = DataFrame(new_df, self.columns, self.index)
-
-        is_bfill = method is not None and method in ['backfill', 'bfill']
-        is_ffill = method is not None and method in ['pad', 'ffill']
-        is_axis_zero = axis is None or axis == 0 or axis == 'index'\
-            or axis == 'rows'
-
-        if is_axis_zero and (is_bfill or is_ffill):
-            def fill_in_part(part, row):
-                return part.fillna(value=row, axis=axis, limit=limit,
-                                   downcast=downcast, **kwargs)
-            last_row_df = None
-            if is_ffill:
-                last_row_df = pd.DataFrame(
-                    [df.iloc[-1, :] for df in ray.get(new_df._df[:-1])]
-                )
-            else:
-                last_row_df = pd.DataFrame(
-                    [df.iloc[0, :] for df in ray.get(new_df._df[1:])]
-                )
-            last_row_df.fillna(value=value, method=method, axis=axis,
-                               inplace=True, limit=limit,
-                               downcast=downcast, **kwargs)
-            if is_ffill:
-                new_df._df[1:] = [
-                    _deploy_func.remote(fill_in_part, new_df._df[i + 1],
-                                        last_row_df.iloc[i, :])
-                    for i in range(len(self._df) - 1)
-                ]
-            else:
-                new_df._df[:-1] = [
-                    _deploy_func.remote(fill_in_part, new_df._df[i],
-                                        last_row_df.iloc[i])
-                    for i in range(len(self._df) - 1)
-                ]
+        if axis == 0:
+            parts = new_obj._col_partitions
+            idx_obj = new_obj._col_index
+        else:
+            parts = new_obj._row_partitions
+            idx_obj = new_obj._row_index
+
+        if isinstance(value, (pd.Series, dict)):
```

Comment:
raise NotImplementedError if value is a dataframe

Reply:
Done above.